planner/core/fragment.go

```diff
@@ -147,13 +158,24 @@ func (m *mppAddr) GetAddress() string {
 
 // for the task without table scan, we construct tasks according to the children's tasks.
 // That's for avoiding assigning to the failed node repeatly. We assumes that the chilren node must be workable.
-func (e *mppTaskGenerator) constructMPPTasksByChildrenTasks(tasks []*kv.MPPTask) []*kv.MPPTask {
+func (e *mppTaskGenerator) constructMPPTasksByChildrenTasks(tasks []*kv.MPPTask, cteProducerTasks []*kv.MPPTask) []*kv.MPPTask {
 	addressMap := make(map[string]struct{})
 	newTasks := make([]*kv.MPPTask, 0, len(tasks))
+	cteAddrMap := make(map[string]struct{})
+	for _, task := range cteProducerTasks {
+		addr := task.Meta.GetAddress()
+		if _, ok := cteAddrMap[addr]; !ok {
+			cteAddrMap[addr] = struct{}{}
+		}
+	}
 	for _, task := range tasks {
 		addr := task.Meta.GetAddress()
 		// for upper fragment, the task num is equal to address num covered by lower tasks
 		_, ok := addressMap[addr]
+		if _, okk := cteAddrMap[addr]; !okk && len(cteAddrMap) > 0 {
```

Comment:
how to understand this？if a children task‘s address is not in the cteProducerAddrs and cteProducerAddrs is not empty, then skip this task?

Reply:
The possible workers are decided from bottom to top. So the address appears in the child fragments must appear in the parent fragments

Reply:
still confused here, take case:

3 tiflash node: A,B,C
one cte producer is from A,B

when shared-cte as one side of join, should left the base table side un-moved. another side is broadcasted.
join
 +-- base table (un-moved)
 +-- receiver2 (cte task from A,B)

Soga: so your code here is meaning that: Consuming-cte means the current OP is join or something, we can always let one side un-moved, just let cte to be as close to the data as possible!?

Reply:
Yes, the current way is easy to implement. But it doesn't use the full nodes of our MPP.
We can support a more enhanced n:m sending strategy to ensure that most data is computed at the local node while we can use the full nodes of our MPP. But it's not contained in this pr.

Reply:
make sense